source/common/grpc/google_async_client_impl.cc

```diff
@@ -0,0 +1,365 @@
+    call_initialized_ = true;
+    rw_->ReadInitialMetadata(tag(Operation::ReadInitialMetadata));
+    writeQueued();
+    break;
+  }
+  case Operation::ReadInitialMetadata: {
+    ASSERT(ok);
+    ASSERT(call_initialized_);
+    rw_->Read(&read_buf_, tag(Operation::Read));
+    Http::HeaderMapPtr initial_metadata = std::make_unique<Http::HeaderMapImpl>();
+    metadataTranslate(ctxt_.GetServerInitialMetadata(), *initial_metadata);
+    callbacks_.onReceiveInitialMetadata(std::move(initial_metadata));
+    break;
+  }
+  case Operation::Write: {
+    ASSERT(ok);
+    write_pending_ = false;
+    write_pending_queue_.pop();
+    writeQueued();
+    break;
+  }
+  case Operation::WriteLast: {
+    ASSERT(ok);
+    write_pending_ = false;
+    break;
+  }
+  case Operation::Read: {
+    ASSERT(ok);
+    std::vector<grpc::Slice> slices;
+    // Assuming this only fails due to OOM.
+    RELEASE_ASSERT(read_buf_.Dump(&slices).ok());
+    // TODO(htuch): As with PendingMessage serialization, the deserialization
+    // here is not optimal, as we are converting between string representation
+    // and have unnecessary copies. We should use ParseFromCodedStream as done
+    // by TensorFlow
+    // https://github.com/tensorflow/tensorflow/blob/f9462e82ac3981d7a3b5bf392477a585fb6e6912/tensorflow/core/distributed_runtime/rpc/grpc_serialization_traits.h#L210
+    // at the cost of some additional implementation complexity.
+    grpc::string buf;
+    buf.reserve(read_buf_.Length());
+    for (const auto& slice : slices) {
+      buf.append(reinterpret_cast<const char*>(slice.begin()), slice.size());
+    }
+    ProtobufTypes::MessagePtr response = callbacks_.createEmptyResponse();
+    if (!response->ParseFromString(buf)) {
+      // This is basically streamError in Grpc::AsyncClientImpl.
+      notifyRemoteClose(Status::GrpcStatus::Internal, nullptr, EMPTY_STRING);
+      resetStream();
+      break;
+    };
+    callbacks_.onReceiveMessageUntyped(std::move(response));
+    rw_->Read(&read_buf_, tag(Operation::Read));
+    break;
+  }
+  case Operation::Finish: {
+    ASSERT(finish_pending_);
+    ENVOY_LOG(debug, "Finish with grpc-status code {}", status_.error_code());
+    Http::HeaderMapPtr trailing_metadata = std::make_unique<Http::HeaderMapImpl>();
+    metadataTranslate(ctxt_.GetServerTrailingMetadata(), *trailing_metadata);
+    const Status::GrpcStatus grpc_status =
+        status_.error_code() <= grpc::StatusCode::DATA_LOSS
+            ? static_cast<Status::GrpcStatus>(status_.error_code())
+            : Status::GrpcStatus::InvalidCode;
+    notifyRemoteClose(grpc_status, std::move(trailing_metadata), status_.error_message());
+    cleanup();
+    break;
+  }
+  default:
+    RELEASE_ASSERT(0);
+  }
+}
+
+void GoogleAsyncStreamImpl::metadataTranslate(
+    const std::multimap<grpc::string_ref, grpc::string_ref>& grpc_metadata,
+    Http::HeaderMap& header_map) {
+  // More painful copying, this time due to the mismatch in header
```

Comment:
Also you need to handle "-bin" suffix metadata, gRPC library does base64 decode / encode for them.

Reply:
Can you explain this one in a bit more detail? It sounds like we need a TODO here, but are there some URLs or GH issues to point at? Isn't -bin a gRPC concept?

Reply:
See https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md:

> Custom-Metadata → Binary-Header / ASCII-Header
> Binary-Header → {Header-Name "-bin" } {base64 encoded value}

gRPC library will do the base64 encode / decode when a header name is '-bin' suffixed. So the behavior doesn't match what Envoy gRPC client does.